File: Etapa3-Construcao/3.1-Stack1_Node_TS/src/domain/entities/Reserva.ts
import { Quarto } from './Quarto.js';
import { Hospede } from './Hospede.js';
import { Disponibilidade } from '../enums/Disponibilidade.js';

// Entidade Reserva controla o vínculo hóspede-quarto.
// Para o escopo sem datas, o estado do quarto determina ocupação.
export class Reserva {
  private _id?: string;
  private _quarto: Quarto;
  private _hospede: Hospede;
  private _criadoEm: Date;
  private _atualizadoEm: Date;
  private _status: 'ATIVA' | 'CANCELADA' = 'ATIVA';

  private constructor(params: { id?: string; quarto: Quarto; hospede: Hospede; criadoEm?: Date }) {
    if (params.quarto.disponibilidade !== Disponibilidade.LIVRE) {
      throw new Error('Quarto indisponível para reserva');
    }
    this._id = params.id;
    this._quarto = params.quarto;
    this._hospede = params.hospede;
    this._criadoEm = params.criadoEm ?? new Date();
    this._atualizadoEm = new Date();
    // Efeito de domínio: marcar quarto como OCUPADO
    this._quarto.alterarDisponibilidade(Disponibilidade.OCUPADO);
  }

  static criar(params: { id?: string; quarto: Quarto; hospede: Hospede; criadoEm?: Date }): Reserva {
    return new Reserva(params);
  }

  trocarHospede(novo: Hospede): void {
    this._hospede = novo;
    this._atualizadoEm = new Date();
  }

  cancelar(): void {
    if (this._status === 'CANCELADA') return;
    this._status = 'CANCELADA';
    this._atualizadoEm = new Date();
    // Ao cancelar, o quarto retorna a LIVRE (simples, sem datas).
    this._quarto.alterarDisponibilidade(Disponibilidade.LIVRE);
  }

  get id(): string | undefined { return this._id; }
  get quarto(): Quarto { return this._quarto; }
  get hospede(): Hospede { return this._hospede; }
  get criadoEm(): Date { return this._criadoEm; }
  get atualizadoEm(): Date { return this._atualizadoEm; }
  get status(): 'ATIVA' | 'CANCELADA' { return this._status; }
}

File: Etapa3-Construcao/3.2-Stack2_Java_Spring/src/main/java/com/hotel/reservas/domain/Reserva.java
package com.hotel.reservas.domain;

import com.hotel.hospedes.domain.Hospede;
import com.hotel.quartos.domain.Disponibilidade;
import com.hotel.quartos.domain.Quarto;
import jakarta.persistence.*;
import lombok.*;

import java.time.OffsetDateTime;

@Entity
@Table(name = "reserva")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Reserva {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    @ManyToOne(optional = false)
    private Quarto quarto;

    @ManyToOne(optional = false)
    private Hospede hospede;

    @Column(nullable = false)
    private OffsetDateTime criadoEm = OffsetDateTime.now();

    @Column(nullable = false)
    private OffsetDateTime atualizadoEm = OffsetDateTime.now();

    @Column(nullable = false)
    private String status = "ATIVA"; // Simples para escopo atual

    @Builder
    private Reserva(Quarto quarto, Hospede hospede) {
        if (quarto.getDisponibilidade() != Disponibilidade.LIVRE) {
            throw new IllegalStateException("Quarto indisponível para reserva");
        }
        this.quarto = quarto;
        this.hospede = hospede;
        // Efeito de domínio: marcar quarto como OCUPADO
        this.quarto.alterarDisponibilidade(Disponibilidade.OCUPADO);
    }

    public void trocarHospede(Hospede novo) {
        this.hospede = novo;
        this.atualizadoEm = OffsetDateTime.now();
    }

    public void cancelar() {
        if ("CANCELADA".equals(this.status)) return;
        this.status = "CANCELADA";
        this.atualizadoEm = OffsetDateTime.now();
        this.quarto.alterarDisponibilidade(Disponibilidade.LIVRE);
    }
}
